### PR TITLE
cutoff option for contact featurizer

### PR DESCRIPTION
**Pull request by brookehus**
- [x] Implement feature / fix bug
 - [x] Add tests
 - [x] Update changelog

Attempts to address #790 although not in a memory-saving way.

Simple code to incorporate mapping contacts to 1/0 if they are under/over a certain cutoff. They are temporarily cast as floats... the code is:
`distances = (distances < self.binary_cutoff).astype(float)`

Obviously can change the type as we want.

Also, I couldn't find in the [docs](http://mdtraj.org/latest/api/generated/mdtraj.compute_contacts.htmll) what the distances units are.

### CODE REVIEW

**Comment by mpharrigan** [comment]
mdtraj always turns everything into nanometers

(http://docs.openmm.org/6.3.0/userguide/theory.html#units for a brief discussion of why this is a good idea (tm))

**Comment by mpharrigan** [comment]
+1 on this pr. some questions:

Why do you cast to float? If you just let it return numpy bools, it would save memory and probably be fine

Add a test(s) :)

We should consider (for usability's sake) having a "suggested" cutoff. I think nate said 8A. There are a couple ways I think we could do this

(a) accept `binary_cutoff='on'`, which means 8 angstroms or whatever we think is best in addition to a float. This is similar to what we do for MSM's `ergodic_cutoff`. We should accept `off` as well if we do this

(b) really I think people expect "contacts" to be binary. We can't change the behavior of this existing featurizer or we might screw up some people's calculations. What we could do is create two featurizers: `BinaryContactsFeaturizer` and `ContactDistanceFeaturizer`, and have the `ContactFeaturizer` be deprecated and say to pick one of those two. If we're patient enough, we can go through a full cycle of [deprecate, remove, re-introduce as binary contacts]

**Comment by cxhernandez** [comment]
> What we could do is create two featurizers: BinaryContactsFeaturizer and ContactDistanceFeaturizer

+1. Also, add a logistic one if it's not too much trouble!

**Comment by brookehus** [comment]
> Why do you cast to float? If you just let it return numpy bools, it would save memory and probably be fine

Is this because tica, clustering, etc. casts to whatever it needs?

And agree, a hybrid, deprecated ContactFeaturizer and then two new featurizers that are unambiguous sound good. I'll incorporate into the PR. @cxhernandez, what's the logistic one supposed to do exactly?

**Comment by cxhernandez** [comment]
Basically the function would just transform the distances using:



where *L* is 1, Xo is the distance cutoff, and k is a tunable parameter which controls the ''softness" of the binning.

**Comment by brookehus** [comment]
What would k be initialized to? Of course it would get thrown into osprey eventually, but is there any precedent?

**Comment by cxhernandez** [comment]
> What would k be initialized to?

Not sure. You would definitely have to play around with it, but my guess would be somewhere around 2 inverse angströms? That way you get from 0 to 1 within half an angstrom. 

Could also be heuristically set as a fraction of the cutoff.

**Comment by brookehus** [comment]
@mpharrigan: in a case where the booleans output from the binary contact featurizer aren't converted to floats, if we don't use tICA and go straight to clustering, e.g. minibatchkmeans, we get an output like:

```
/Users/brooke/anaconda/lib/python2.7/site-packages/sklearn/cluster/k_means_.py:1433: RuntimeWarning: Got data type bool, converted to float to avoid overflows
  X = self._check_test_data(X)
```

Therefore, to me, it seems preferable to convert to float within the featurizer. What do you think?

**Comment by cxhernandez** [comment]
Woud `uint8 ` be a compromise?

**Comment by brookehus** [comment]
I think it wants a float. `float16`?

**Comment by mpharrigan** [comment]
I don't see the problem. kmeans apparently wants float for a legitimate reason, and they cast to float. no problem.

File: msmbuilder/featurizer/featurizer.py [hunk 1]
```diff
@@ -851,8 +851,132 @@ def __init__(self, mode='residue', **kwargs):
     def partial_transform(self, traj):
         return md.shrake_rupley(traj, mode=self.mode, **self.kwargs)
 
-
+# BEH: normal
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: ?

File: msmbuilder/featurizer/featurizer.py [hunk 2]
```diff
@@ -851,8 +851,132 @@ def __init__(self, mode='residue', **kwargs):
...
 class ContactFeaturizer(Featurizer):
+    """Deprecated. Please use BinaryContactFeaturizer or
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: The intro sentence should be one line

**Comment by mpharrigan** [comment]
Looks ok to me.

1. can we reduce code duplication?

2. can you put the new argument (binary_cutoff) last, to retain backwards compatibility when using positional arguments

**Comment by brookehus** [comment]
@mpharrigan @cxhernandez redone with a base class for distances and two classes that inherit it (binary and logistic), with a test, and with a few lines in some clustering codes to make sure things are cast as floats first, since that is what libdistance requires.

File: msmbuilder/commands/featurizer.py [hunk 3]
```diff
@@ -169,6 +170,26 @@ def _contacts_type(self, val):
         else:
             return np.loadtxt(val, dtype=int, ndmin=2)
 
+class BinaryContactFeaturizerCommand(FeaturizerCommand):
+    _concrete = True
+    klass = BinaryContactFeaturizer
+
+    def _contacts_type(self, val):
+        if val is 'all':
```
Comment (mpharrigan) [inline msmbuilder/commands/featurizer.py]: this needs to use `==`

Comment (mpharrigan) [inline msmbuilder/commands/featurizer.py]: http://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce

File: msmbuilder/featurizer/featurizer.py [hunk 4]
```diff
@@ -964,6 +967,106 @@ def describe_features(self, traj):
...
+            'closest-heavy' : distance is the closest distance between
+                any two non-hydrogen atoms in the residues
+    ignore_nonprotein : bool
+        When using `contact==all`, don't compute contacts between
+        "residues" which are not protein (i.e. do not contain an alpha
+        carbon).
+    center : float, default=0.8
+        Determines the midpoint of the sigmoid, x_0
+    steepness : float, default=2
+        Determines the steepness of the logistic curve, k
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: These parameters have units I think. Might want to mention them here

Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: `nm` and `1/nm`, resp

**Comment by mpharrigan** [comment]
Test failures seem relevant

File: msmbuilder/featurizer/featurizer.py (hunk 4 above)
Comment (cxhernandez) [inline msmbuilder/featurizer/featurizer.py]: I think steepness should be 20 `1/nm`

File: msmbuilder/commands/featurizer.py (hunk 3 above)
Comment (brookehus) [inline msmbuilder/commands/featurizer.py]: @mpharrigan the real lesson here is that i shouldn't copy and paste code

File: msmbuilder/featurizer/featurizer.py [hunk 5]
```diff
@@ -964,6 +967,110 @@ def describe_features(self, traj):
...
+        distances larger than CENTER will return values smaller than 0.5.
+    steepness : float, default=20
+        Determines the steepness of the logistic curve, [1/nm]. Small 
+        and large distances will approach ouput values of 1 and 0,
+        respectively, more quickly.
+    """
+
+
+    def __init__(self, contacts='all', scheme='closest-heavy', ignore_nonprotein=True,
+                                                            center=0.8, steepness=2):
```
Comment (cxhernandez) [inline msmbuilder/featurizer/featurizer.py]: change the default value of steepness

File: msmbuilder/tests/test_contactfeaturizers.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def test_binaries():
+    dataset = fetch_fs_peptide()
+    trajectories = dataset["trajectories"]
+    binarycontactfeaturizer = BinaryContactFeaturizer()
+    binaries = binarycontactfeaturizer.transform([trajectories[0]])
+
+    assert binaries[0].shape[1] == 171
+    assert sum(sum(binaries[0])) <= binaries[0].shape[0]*binaries[0].shape[1]
```
Comment (cxhernandez) [inline msmbuilder/tests/test_contactfeaturizers.py]: You could just use `np.sum` once, but what might be more appropriate is to check that the `dtype` of this numpy object is `bool`?

File: msmbuilder/tests/test_contactfeaturizers.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def test_binaries_inf_cutoff():
```
Comment (cxhernandez) [inline msmbuilder/tests/test_contactfeaturizers.py]: Could be worth adding a `test_binaries_zero_cutoff` as well

File: msmbuilder/tests/test_contactfeaturizers.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    steepness = 10*np.random.randn()
+    center = np.random.randn()
+    binarycontactfeaturizer = BinaryContactFeaturizer(cutoff=center)
+    binaries = binarycontactfeaturizer.transform([trajectories[0]])
+    logisticcontactfeaturizer = LogisticContactFeaturizer(center=center, steepness=steepness)
+    logistics = logisticcontactfeaturizer.transform([trajectories[0]])
+
+    # This checks that no distances that are larger than the center are logistically
+    # transformed such that they are less than 1/2
+    assert sum(sum((logistics[0]*binaries[0] < 0.5)*(logistics[0]*binaries[0] > 0.0))) == 0
```
Comment (cxhernandez) [inline msmbuilder/tests/test_contactfeaturizers.py]:
```python
assert all((logistics[0].ravel() >= 0.5) - (binaries[0].ravel() == True) == 0)
```

File: msmbuilder/tests/test_contactfeaturizers.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    test = np.log(1.0/y-1)/steepness + center
+
+    np.testing.assert_almost_equal(x, test, decimal=2)
+
+
+def test_binary_to_logistics():
+    dataset = fetch_fs_peptide()
+    trajectories = dataset["trajectories"]
+    steepness = 10*np.random.randn()
+    center = np.random.randn()
```
Comment (cxhernandez) [inline msmbuilder/tests/test_contactfeaturizers.py]: This test is throwing an error on travis-ci. My guess is that this line and the one before it are spitting out negative values.

Comment (mpharrigan) [inline msmbuilder/tests/test_contactfeaturizers.py]: better yet, add something to the init method to assert that these are positive and give the user a helpful error message if not

File: msmbuilder/cluster/kcenters.py [hunk 10]
```diff
@@ -77,6 +77,9 @@ def __init__(self, n_clusters=8, metric='euclidean', random_state=None):
         self.random_state = random_state
 
     def fit(self, X, y=None):
+        if isinstance(X, np.ndarray):
+            if X.dtype != 'float32' and X.dtype != 'float64':
```
Comment (cxhernandez) [inline msmbuilder/cluster/kcenters.py]: Not sure the check is really necessary. I would just go ahead and cast it:

```python
X = X.astype(float)
```

**Comment by cxhernandez** [comment]
aside from my really minor comment above, this lgtm

**Comment by brookehus** [comment]
I tried that, but some tests failed. I think libdistance requires that when there are multiple inputs (X and y) that they be the same type of float (either 32 or 64).

**Comment by mpharrigan** [comment]
Maybe the check would be more idiomatic as
```python
if not (dype==float32 or dtype==float64)
```

But this looks good. Update changelog!

**Comment by brookehus** [comment]
yuck why did the changelog get weird here??

**Comment by mpharrigan** [comment]
I bet there's not supposed to be a space before the `-`s. I'll fix that in a follow-up pr

**Comment by mpharrigan** [comment]
I restarted the travis failure, which is one I haven't seen before and hopefully a one-off thing. Windows tests don't even build. I'll have to boot into windows one of these days and see if I can fix things

**Comment by mpharrigan** [comment]
The py35 tests keep getting killed?

Also these new tests take over 20 seconds. I'd like to stop adding new tests that are "red" for time

**Comment by mpharrigan** [comment]
Actually they're like 6 seconds on the other travis runs. Why is it slower on other travis?

**Comment by cxhernandez** [comment]
bump @mpharrigan

**Comment by mpharrigan** [comment]
I'll restart the tests but my issues still stand

**Comment by brookehus** [comment]
What are your issues specifically? I can't tell from your comments above what I need to do.

**Comment by mpharrigan** [comment]
The tests fail and seem to hang in the newly added tests/code

**Comment by brookehus** [comment]
I have no idea about the travis error message for the third build, `No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.`, and also have no idea why `test_binary_to_logistics` takes SO long. It doesn't take that long on my laptop. Do you think I should just get rid of it?

**Comment by mpharrigan** [comment]
Let's not delete it just yet. It might be indicative of a real problem.

**Comment by brookehus** [comment]
Okay, well given that it runs fine and in <10s on my machine, how do you suggest gaining further insight?

**Comment by cxhernandez** [comment]
`msmbuilder.tests.test_contactfeaturizers.test_binary_to_logistics ... (xdrfile error) Undocumented error 3ERROR`

could this be something weird with mdtraj then?

**Comment by rmcgibbo** [comment]
Is the XTC file corrupted? That's the kind of error I would expect mdtraj to throw in that situation.

**Comment by cxhernandez** [comment]
Those same XTCs working for the other featurizer tests though, right?

**Comment by mpharrigan** [comment]
@brookehus Do you want to try rebasing this and letting the tests re-run? Maybe it will magically work and we can get this in to v3.5

**Comment by cxhernandez** [comment]
Still something weird going on in py3.4... but doesn't seem to be a result of this PR

**Comment by mpharrigan** [comment]
Wanna rebase now the #830 is merged?

**Comment by cxhernandez** [comment]
This looks good now!

**Comment by dr-nate** [comment]
Yayyy!

**Comment by brookehus** [comment]
tbt to the never-ending #798 thread

**Pull request merged**